packages/edit-site/src/components/sidebar/color-panel.js

```diff
@@ -0,0 +1,112 @@
+/**
+ * WordPress dependencies
+ */
+import { __experimentalPanelColorGradientSettings as PanelColorGradientSettings } from '@wordpress/block-editor';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import {
+	BACKGROUND_COLOR,
+	LINK_COLOR,
+	TEXT_COLOR,
+	GRADIENT_COLOR,
+} from '../editor/utils';
+
+export default ( {
+	context: { supports, name },
+	getProperty,
+	setProperty,
+} ) => {
+	if (
+		! supports.includes( TEXT_COLOR ) &&
+		! supports.includes( BACKGROUND_COLOR ) &&
+		! supports.includes( GRADIENT_COLOR ) &&
+		! supports.includes( LINK_COLOR )
+	) {
+		return null;
+	}
+
+	const settings = [];
+
+	if ( supports.includes( TEXT_COLOR ) ) {
+		settings.push( {
+			colorValue: getProperty( name, 'color.text' ),
+			onColorChange: ( value ) =>
+				setProperty( name, { 'color.text': value } ),
+			label: __( 'Text color' ),
+		} );
+	}
+
+	/*
+	 * We want to send the entities API both colors
+	 * in a single request. This is to avod race conditions
+	 * that override the previous callback.
+	 */
+	let setBackground;
+	let backgroundSettings;
+	const backgroundPromise = new Promise(
+		( resolve ) => ( setBackground = ( value ) => resolve( value ) )
+	);
+	let setGradient;
+	let gradientSettings;
+	const gradientPromise = new Promise(
+		( resolve ) => ( setGradient = ( value ) => resolve( value ) )
+	);
+	Promise.all( [ backgroundPromise, gradientPromise ] ).then( ( values ) => {
```

Comment:
Understanding how the values are set is really complex, and I guess we should try to improve this in a follow-up PR.
Calling a Promise.all directly inside the render may make the render impure and go against the react paradigm no? Because executing a promise is like a side-effect of the render.

Reply:
Yeah, at first glance this looks like something that belongs in a `useEffect`.